Behaviors/GroupBoxBehaviors.cs

```diff
@@ -22,69 +22,58 @@ private static void OnHideOnChildrenNotVisibleChanged(DependencyObject obj, Depe
+        {
+            var multiBinding = new MultiBinding
+            {
+                Converter = new MultiVisibilityToVisibilityConverter()
+            };
+
+            if (newBehaviorPropertyValue)
+            {
+                foreach (var child in FindVisualChildren<Button>(groupBox))
                 {
-                    BindingOperations.ClearBinding(groupBox, UIElement.VisibilityProperty);
+                    multiBinding.Bindings.Add(new Binding("Visibility") { Mode = BindingMode.OneWay, Source = child });
                 }
-                
+                BindingOperations.SetBinding(groupBox, UIElement.VisibilityProperty, multiBinding);
             }
             else
             {
-                groupBox.Loaded += (sender, eventArgs) =>
-                {
-                    var multiBinding = new MultiBinding
-                    {
-                        Converter = new MultiVisibilityToVisibilityConverter()
-                    };
-
-                    if ((bool)args.NewValue)
-                    {
-                        foreach (var child in FindVisualChildren<Button>(groupBox))
-                        {
-                            multiBinding.Bindings.Add(new Binding("Visibility") { Mode = BindingMode.OneWay, Source = child });
-                        }
-                        BindingOperations.SetBinding(groupBox, UIElement.VisibilityProperty, multiBinding);
-                    }
-                    else
-                    {
-                        BindingOperations.ClearBinding(groupBox, UIElement.VisibilityProperty);
-                    }
-                };
+                BindingOperations.ClearBinding(groupBox, UIElement.VisibilityProperty);
             }
-
         }
 
         public static IEnumerable<T> FindVisualChildren<T>(DependencyObject depObj) where T : DependencyObject
         {
+            List<T> childElements = new List<T>();
             if (depObj != null)
             {
                 for (int i = 0; i < VisualTreeHelper.GetChildrenCount(depObj); i++)
                 {
                     DependencyObject child = VisualTreeHelper.GetChild(depObj, i);
+                    var childVisibility = (Visibility)child.GetValue(UIElement.VisibilityProperty);
+                    
                     if (child != null && child is T)
                     {
-                        yield return (T)child;
+                        childElements.Add((T)child);
                     }
 
-                    foreach (T childOfChild in FindVisualChildren<T>(child))
+                    if (VisualTreeHelper.GetChildrenCount(child) > 0 && childVisibility == Visibility.Visible)
                     {
-                        yield return childOfChild;
+                        childElements.AddRange(FindVisualChildren<T>(child));
                     }
                 }
             }
+
+            return childElements;
         }
     }
 }
```

Comment:
Come to think of it... We can't use this approach, because any element can change it's visibility at any time, we can't only bind to the ones that are visible at the time of load. I suggest using recursion only to hunt for buttons (if possible, only hunt for the ones that have a workbook command associated with them). Then bind to them. Alternatively, you could add another attached property to mark the buttons (or any other control we care about) and look up from that one (or down from the groupbox one).

Reply:
@anakic I updated the approach. Tl;dr version: The buttons have a behavior where they are provided a parent element. Then they "register" themselves to the parent and then the parent binds to the children properties.

Reply:
Question to think about: Is this solution applicable outside of QueryStorm? Would these behaviors make sense for someone else in a different project (without too much explanation)? I.e. should this be in the main QueryStorm project because the functionality is specific to QS, or is this other people can reuse in other projects?

Reply:
I think this could be reusable. I don't see why not, the behavior is documented and someone can easily extend it to support more than just the Visibility parent-children binding. And to avoid repeating the same behavior property on each child element, they could in turn create a style (either new one or based on an existing style) where it's set only once and applied  to each child.

Reply:
When I was searching for a solution, all answers I found had one thing in common - traversing the tree (be it a visual or a logical tree). This is more rigid as you specifically state in the layout - hey, I want this child to impact the properties of the specified parent => no crazy edge cases in recursion, in fact, no recursion and no head banging why some event is being triggered or not.